python/ray/serve/_private/application_state.py

```diff
@@ -334,50 +403,89 @@ def _determine_app_status(self) -> Tuple[ApplicationStatus, str]:
         else:
             return ApplicationStatus.DEPLOYING, ""
 
-    def _reconcile_deploy_obj_ref(self) -> Tuple[BuildAppStatus, str]:
-        """Reconcile the in-progress deploy task.
+    def _start_or_reconcile_build_app_task(
+        self, code_version: str
+    ) -> Tuple[List, BuildAppStatus, str]:
+        """Start or reconcile the in-progress build task.
 
-        Resets the self._deploy_obj_ref if it finished, regardless of
-        whether it finished successfully.
+        If there is no in-progress build task with the correct code
+        version, start a new one. Otherwise, reconcile the in-progress
+        build task (resets self._build_app_task_info when task finishes,
+        regardless of whether it finished successfully)
 
         Returns:
+            Deploy arguments (List[Dict]):
+                The deploy arguments returned from the build app task.
             Status (BuildAppStatus):
                 SUCCEEDED: task finished successfully.
                 FAILED: an error occurred during execution of build app task
                 IN_PROGRESS: task hasn't finished yet.
             Error message (str):
                 Non-empty string if status is DEPLOY_FAILED or UNHEALTHY
         """
-        if check_obj_ref_ready_nowait(self._deploy_obj_ref):
-            deploy_obj_ref, self._deploy_obj_ref = self._deploy_obj_ref, None
+        if (
+            self._build_app_task_info is None
+            or self._build_app_task_info.code_version != code_version
+        ):
+            # If there is an in progress build task, cancel it.
+            if self._build_app_task_info:
+                logger.info(
+                    f'Received new config for application "{self._name}". '
+                    "Cancelling previous request."
+                )
+                ray.cancel(self._build_app_task_info.obj_ref)
+
+            # Halt reconciliation of target deployments
+            self._set_target_state_deployment_infos(None)
+
+            # Kick off new build app task
+            logger.info(
+                f"Starting build_serve_application task for application {self._name}."
+            )
+            build_app_obj_ref = build_serve_application.options(
+                runtime_env=self._target_state.config.runtime_env
+            ).remote(
+                self._target_state.config.import_path,
+                self._target_state.config.runtime_env,
+                self._target_state.config.deployment_names,
+                code_version,
+                self._target_state.config.name,
+                self._target_state.config.args,
+            )
+            self._build_app_task_info = BuildAppTaskInfo(
+                build_app_obj_ref, code_version
+            )
+        elif check_obj_ref_ready_nowait(self._build_app_task_info.obj_ref):
+            build_app_obj_ref = self._build_app_task_info.obj_ref
+            self._build_app_task_info = None
```

Comment:
That's gated by line 524 - once the task finishes and the deploy args are fetched, the deploy args + the code version are applied to the target state through `apply_deployment_args`. Then in the next iteration, even though `_build_app_task_info` is None, since the current code version matches that of the target config, `_start_or_reconcile_build_app_task` isn't called.

It is a bit confusing because we need to handle the target config being out of sync with the rest of the target state, and correctly detect that to bring the rest of the target state up to date.

Reply:
Got it. Without being as deep in the code as you I think it would be slightly preferable to make this function itself handle the case where the build task has already been run, but if this structure makes more sense to you then I'm OK with it.

Regardless, please add a comment about this assumption to the docstring of `_start_or_reconcile_build_app_task` and a short comment to the callsite saying that we should skip calling it when the code version already matches.

Reply:
Ah I see! I've changed it so that the function is always called but will return a `NO_TASK_STARTED` status if there is no in progress task (nor does one need to be started).